### ISSUE TITLE
[Performance]: Extremely slow FP32 inference on C3 (Sapphire Rapids) compared to N2 (Ice Lake)

### ISSUE THREAD

**Issue by deepanshuhardaha** [openvinotoolkit/openvino#32405]
### OpenVINO Version

openvino/model_server:2025.3

### Operating System

Ubuntu 22.04 (LTS)

### Device used for inference

CPU

### OpenVINO installation

Docker

### Programming Language

Python

### Hardware Architecture

x86 (64 bits)

### Model used

Custom TensorFlow Keras dense ranking model (converted via Model Optimizer to OpenVINO IR format)

### Model quantization

No

### Target Platform

Architecture:                x86_64
  CPU op-mode(s):            32-bit, 64-bit
  Address sizes:             52 bits physical, 57 bits virtual
  Byte Order:                Little Endian
CPU(s):                      88
  On-line CPU(s) list:       0-87
Vendor ID:                   GenuineIntel
  Model name:                Intel(R) Xeon(R) Platinum 8481C CPU @ 2.70GHz
    CPU family:              6
    Model:                   143
    Thread(s) per core:      2
    Core(s) per socket:      44
    Socket(s):               1
    Stepping:                8
    BogoMIPS:                5399.99
    Flags:                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopo
                             logy nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefe
                             tch ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd 
                             sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 arat avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_
                             vpopcntdq rdpid cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk amx_bf16 avx512_fp16 amx_tile amx_int8 arch_capabilities
Virtualization features:     
  Hypervisor vendor:         KVM
  Virtualization type:       full
Caches (sum of all):         
  L1d:                       2.1 MiB (44 instances)
  L1i:                       1.4 MiB (44 instances)
  L2:                        88 MiB (44 instances)
  L3:                        105 MiB (1 instance)
NUMA:                        
  NUMA node(s):              2
  NUMA node0 CPU(s):         0-21,44-65
  NUMA node1 CPU(s):         22-43,66-87
Vulnerabilities:             
  Gather data sampling:      Not affected
  Indirect target selection: Mitigation; Aligned branch/return thunks
  Itlb multihit:             Not affected
  L1tf:                      Not affected
  Mds:                       Not affected
  Meltdown:                  Not affected
  Mmio stale data:           Not affected
  Reg file data sampling:    Not affected
  Retbleed:                  Not affected
  Spec rstack overflow:      Not affected
  Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB disabled; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
  Srbds:                     Not affected
  Tsx async abort:           Not affected

### Performance issue description

We are observing extremely high latency and poor throughput when using `INFERENCE_PRECISION_HINT="f32"` on OVMS running on N4/C3 (Sapphire Rapids) CPUs.

With `INFERENCE_PRECISION_HINT="bf16"`, we achieve expected throughput, but the accuracy drift compared to the baseline (N2) is significant for us, which we want to avoid. When we force `INFERENCE_PRECISION_HINT="f32",` latency increases 3–4×, and throughput drops by ~60%, even though CPU utilization increases by ~60%.

- Using `DNNL_VERBOSE=1`, we confirmed that the FP32 path uses AVX512.
- On N2, FP32 throughput and latency are stable and ~2× faster than the same model on N4/C3 with FP32.
- Performance comparison with identical batch size and pod resources
  - **Performance on C3**: RPS: 101, p50: 50 ms, p95: 104 ms, CPU utilization: 56%
  - **Performance on C3 with `INFERENCE_PRECISION_HINT="f32"`**: RPS: 69, p50: 172 ms, p95: 256 ms, CPU utilization: 89.1%
  - **Performance on N2**: RPS: 51, p50: 101 ms, p95: 194 ms, CPU utilization: 47.9%
- Tested using latest `OpenVINO / model_server:2025.3`, issue persists.

Is such a significant drop in performance expected when using `INFERENCE_PRECISION_HINT="f32"` or am I missing something?

### Step-by-step reproduction

1. Convert a custom TensorFlow Keras model to OpenVINO IR (FP32) and save it to the specified path using the code below
  ```
  converted_model = ov.convert_model(
      tfModel,
      input=[-1, 2708],
      share_weights=False,
      verbose=True,
  )
  
  ov.save_model(
      converted_model,
      save_path,
      compress_to_fp16=False,
  )
  ```
2. Deploy the saved model via openvino/model_server:2025.3:
```
docker run -d \
  -v /models:/models \
  -p 9000:9000 -p 8000:8000 \
  openvino/model_server:2025.3 \
  --model_path /models/my_model \
  --port 9000 --rest_port 8000 \
  --plugin_config '{"INFERENCE_PRECISION_HINT":"f32"}'
```

### Issue submission checklist

- [x] I'm reporting a performance issue. It's not a question.
- [x] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [x] There is reproducer code and related data files such as images, videos, models, etc.

**Comment by ABL4Z3** [comment]
Hi @deepanshuhardaha,

This is an excellent and very detailed performance report. Thank you for providing all the necessary information to reproduce and understand the issue.

Your analysis is correct—you've identified a significant performance regression for the FP32 inference path on the new Sapphire Rapids (C3) architecture compared to Ice Lake (N2).

The Technical Reason 
Sapphire Rapids CPUs introduce new hardware units called AMX (Advanced Matrix Extensions), which provide a massive speedup for lower-precision formats like BF16. By default, OpenVINO smartly uses this path, which is why you see great performance initially.

When you force INFERENCE_PRECISION_HINT="f32", OpenVINO must fall back to using the AVX512 instruction set. Your report strongly suggests that this FP32 code path is not as well-optimized on this new hardware as it was on the previous generation, leading to the performance drop you're seeing. This is valuable feedback for the development team.

Recommended Solutions 
While the team investigates the FP32 performance, here are a few workarounds you can try to get the best of both worlds (performance and accuracy) on the new C3 instances:

Quantization-Aware Training (QAT) - (Most Recommended) The most promising solution is to retrain your model using QAT. This technique can help you regain most of the accuracy lost when converting to BF16, allowing you to take full advantage of the high-speed AMX hardware. Since your model is in TensorFlow, you can use the [TensorFlow Model Optimization Toolkit](https://www.tensorflow.org/model_optimization/guide/quantization/training) for this.

Experiment with CPU Plugin Settings You can try to fine-tune the execution of the FP32 model by adjusting OpenVINO's CPU plugin configurations. For example, manually setting the number of performance streams via ov::num_streams (CPU_THROUGHPUT_STREAMS) might change the behavior. It's not guaranteed to fix the core regression, but it's worth experimenting with.

In summary, this seems to be a genuine software optimization issue for FP32 on this new hardware. For an immediate fix, QAT is likely your best path forward.

**Comment by deepanshuhardaha** [comment]
Hi @ABL4Z3 

I’ve already tried various OpenVINO configs, including `PERFORMANCE_HINT`, `NUM_STREAMS`, `EXECUTION_MODE_HINT`, `ENABLE_CPU_PINNING`, `ENABLE_HYPER_THREADING`, `nireq`, `grpc_workers`, etc., as well as DNN-related settings like `ONEDNN_MAX_CPU_ISA`, but nothing seems to work.

Only adding `"ENABLE_HYPER_THREADING": "NO"` along with `"INFERENCE_PRECISION_HINT": "f32"` slightly improved performance, but it still remained poor. Even after deploying this config in production, the performance was still significantly worse.

I haven’t yet tried selective FP32 execution using NNCF or TensorFlow’s mixed-precision quantization approach ([TensorFlow Model Optimization Guide](https://www.tensorflow.org/model_optimization/guide/quantization/training)), as you suggested.
However, when I ran TensorFlow Serving on Sapphire Rapids (without any config tuning), it worked fine and achieved 2× the performance of Ice Lake (N2) with no accuracy loss. However, TFServing performance on Sapphire Rapids is about 60% lower than OVMS when OVMS runs with BF16 precision.

Performance summary:

**TFServing on N2:** RPS: 27.7, p50: 42.2 ms, p95: 64.9 ms, CPU utilization: 77%
**TFServing on N4:** RPS: 44.2, p50: 25.2 ms, p95: 36.3 ms, CPU utilization: 73%
**OVMS on N4:** RPS: 101, p50: 50 ms, p95: 104 ms, CPU utilization: 56%

Any idea what TF Serving is doing to avoid accuracy loss while still maintaining performance — and whether there’s a way to achieve the same with OVMS?

**Comment by ABL4Z3** [comment]
Thank you for this excellent follow-up and for providing the TFServing benchmark data. That's a critical clue, and it actually helps confirm the root cause.

Your analysis is spot on, and this new data points to a very specific interaction between the model, OpenVINO, and the hardware.

Here’s what is almost certainly happening based on your new findings:

Answering Your Question: What is TFServing Doing?

TensorFlow Serving, when running on Intel hardware (especially Sapphire Rapids), also uses the oneDNN library (just like OpenVINO). However, it is likely not running in pure FP32 mode, even though your model is FP32.

By default, oneDNN on a platform with AMX (Advanced Matrix Extensions) will often enable a setting called fpmath_mode=bf16. This means that even for FP32 models, it performs compute-heavy operations (like matrix multiplications) by:

Transparently and automatically downcasting the FP32 inputs/weights to BF16.

Executing the operation using the extremely fast AMX hardware units.

Casting the BF16 result back up to FP32.

This "mixed precision" approach explains all of your observations:

Why TFServing is 2x faster on N4 (C3): It's using the AMX hardware (the BF16 path), which is massively faster than the AVX512 (FP32) path.

Why TFServing has no accuracy loss: This on-the-fly, internal-only conversion is far less aggressive than a full, static model quantization. For many models, the accuracy impact is negligible.

What is OpenVINO Doing?

Default (Fast, Inaccurate): When you run OVMS without any hints, it also tries to use this fast AMX/BF16 path. However, your model appears to be sensitive to OpenVINO's specific implementation of this, causing the "significant accuracy drift" you want to avoid.

INFERENCE_PRECISION_HINT="f32" (Slow, Accurate): When you add this hint, you are explicitly forbidding OpenVINO from using the fast AMX/BF16 path. It is respecting your command and falling back to the "pure" FP32 (AVX512) path, which (as we've established) is unfortunately not well-optimized on this new architecture, leading to the terrible performance.

The New Core Problem & Solution

The issue isn't that FP32 is inherently slow on C3. The issue is that your model is sensitive to the fast BF16/AMX path when used via OpenVINO.

TFServing's performance proves the hardware is capable of high speed with good accuracy. We just need to make your model "friendly" to this fast path within OpenVINO.

Recommended Next Steps:

Confirmation (Optional): If you can, try running your TFServing container again, but add the environment variable ONEDNN_VERBOSE=1. If you see kernel names like amx_bf16 or brgemm:bf16 in the verbose logs, this will confirm 100% that it's using the AMX path.

Solution (Re-iterate): This new data points even more strongly to Quantization-Aware Training (QAT) as the correct production solution. Your model needs to be made "BF16-aware" to run robustly on this hardware's fastest path. By re-training with QAT, you will get the accuracy of FP32 while unlocking the BF16/AMX performance that you see in both TFServing and default OVMS.

Alternative (NNCF): As you mentioned, your other path is using NNCF for selective precision. You could try to manually keep only the most "sensitive" layers in FP32 and move the rest to BF16. This is a good alternative if QAT is too difficult to implement.

In summary: TFServing is likely "cheating" by using automatic mixed precision. To get the same benefit in OpenVINO, you need to prepare your model for it using QAT.
